mipengine/algorithms/pca.py

```diff
@@ -0,0 +1,123 @@
+import json
+from typing import TypeVar, List
+
+import numpy
+from pydantic import BaseModel
+
+from mipengine.udfgen.udfgenerator import (
+    udf,
+    transfer,
+    state,
+    relation,
+    make_unique_func_name,
+    merge_transfer,
+)
+
+
+class PCAResult(BaseModel):
+    title: str
+    n_obs: int
+    eigenvalues: List[float]
+    eigenvectors: List[List[float]]
+
+
+def run(algo_interface):
+    local_run = algo_interface.run_udf_on_local_nodes
+    global_run = algo_interface.run_udf_on_global_node
+
+    X_relation = algo_interface.initial_view_tables["x"]
+
+    local_transfers = local_run(
+        func_name=make_unique_func_name(local1),
+        keyword_args={"x": X_relation},
+        share_to_global=[True],
```

Comment:
len(share_to_global) is always supposed to be equal to the length of the keyword_args or the positional_args. The AlgorithmExecutor checks that and throws an exception if the equality's not standing. Since keyword_args and positional_args are a Dict or a List, respectively, so I think it is more intuitive that the share_to_global is always a List[bool]

Reply:
`len(share_to_global)` should be equal to the number if items returned by the UDF, not to the number of arguments. A UDF can return a single (non-tuple) item, or a tuple with multiple items. In the first case we accept `share_to_global` to be of type `bool` and in the second it should be of type `List[bool]`. I think this is very consistent.